### DISCUSSION TITLE
代码里如何显示候选词小窗？

### DISCUSSION THREAD

**Discussion by guyuming76** [fcitx/fcitx5#368] [Q&A]
我遇到了 wps 里输入法选词小窗口不显示的问题， 我的环境是 kwin_wayland + xwayland, 并且xorg-server 包没装。我发觉可能是 装上 xorg-server 后问题就解决了：   https://www.zhihu.com/question/490979954/answer/2181078176

我现在在另一台机器上能重现这个问题，在尝试安装 xorg-server 前，我想从架构原理上理解这个问题，并能够结合调试代码理解现象。

我读了 https://www.csslayer.info/wordpress/fcitx-dev/fcitx-5-key-event-handling/ ，但最后一段讲classicui啥的似乎不够详细，我参照文章用cgdb调试，但也仅限于执行到 instance.cpp 里那些匿名函数写的eventhandler. 针对我遇到的问题，该如何设置断点？

**Comment by wengxt** [comment]
这个比较复杂，现在根据环境的不同有非常多可能的路线，Xwayland 的话就和普通的 X 没有区别，你装 gentoo 的话我不知道是不是你装 xorg-server 才能用 xwayland。可能一开始根本你就没用成 xwayland

在wayland下，fcitx 不会用 wayland 给 dbusfrontend 的程序画界面，它只会选择 1、用X画，2、让程序自己画（fcitx5-qt / gtk），而 wps 不是用 fcitx 的 im module 实现，可能根本没有实现这个功能。不用 wayland 画界面因为没法绝对定位窗口位置，画了也没有任何意义

你可以用这个命令看看程序运行在什么模式下面
 dbus-send --print-reply=literal --dest=org.fcitx.Fcitx5 /controller org.fcitx.Fcitx.Controller1.DebugInfo

**Reply by guyuming76** [reply]
多谢，下面是我的返回结果：

ggy@gymGentoo /usr/src/debug/app-i18n $ dbus-send --print-reply=literal --dest=org.fcitx.Fcitx5 /controller org.fcitx.Fcitx.Controller1.DebugInfo
   Group [wayland:] has 9 InputContext(s)
  IC [08d74c758aed4d08ab8066b8ee36155a] program:firefox frontend:dbus cap:e001000012 focus:0
  IC [85f33d15d29a4035b0a6984dd98075e4] program:firefox frontend:dbus cap:e001000012 focus:0
  IC [396cb6b9cf34403f9eb1ac5814377e9b] program:spectacle frontend:dbus cap:e001800072 focus:0
  IC [5558de70b6f74dd79fc0df4133ddeae2] program:firefox frontend:dbus cap:e001000012 focus:0
  IC [f4767897ae444a18b42c5a74c9827cd9] program:konsole frontend:dbus cap:e001800072 focus:0
  IC [f56fe0df22814fd3bd051276edf06e8b] program:konsole frontend:dbus cap:e001800072 focus:1
  IC [b40efb63afc34a39a200c5d0910db994] program:firefox frontend:dbus cap:e001000052 focus:0
  IC [59440f991cf94d709723755012aaa23b] program:plasmashell frontend:dbus cap:e001800072 focus:0
  IC [fa63ccec6cc142eea88934ce8f1ac0af] program:plasmashell frontend:dbus cap:e001800072 focus:0
Input Context without group
  IC [705a4fdfa3d34bf79b8f38e051859ef8] program:wps frontend:dbus focus:0
  IC [4a487411b68b48ceb8e8b6c68bc80b86] program:wps frontend:dbus focus:0

**Comment by guyuming76** [comment]
gentoo 装系统一开始我是装 xorg-server 的，后来用wayland, 我就把 xorg-server 相关包给卸了，但是 xwayland 是出现在进程树里面的：

![Screenshot_20211031_071546](https://user-images.githubusercontent.com/12045548/139561065-688bcd4a-0655-4009-b17b-cde9ec79e02c.png)

**Reply by wengxt** [reply]
你的图片坏了

**Reply by guyuming76** [reply]
图片重新上传了, WPS里能输入汉字，但没有选词小窗，只能按空格输入第一个词组。

**Reply by wengxt** [reply]
Input Context without group wps 没有获得分组这个地方比较奇怪，因为我大概看了一下最新的 wps 他们看起来已经切换到了 fcitx 实现的 qt im module。假定是我的实现的话，WPS 应该声明它自己是用 X11

调用大概会类似

method call time=1635637107.141815 sender=:1.200 -> destination=:1.214 serial=82 path=/org/freedesktop/portal/inputmethod; interface=org.fcitx.Fcitx.InputMethod1; member=CreateInputContext
   array [
      struct {
         string "program"
         string "wps"
      }
      struct {
         string "display"
         string "x11:"
      }
   ]

那问题就是 fcitx5 自己没有 X11。我觉得问题是你没有开某个 fcitx5 的 USE

如果 Fcitx5 连接到了 X11，应该会显示
Group [x11::1] has 0 InputContext(s)

**Reply by guyuming76** [reply]
谢谢，gentoo 里 fcitx5 加 X USE flag 重新编译安装启动就好了, 我机器 make.conf 里全局范围是减X的。xorg-server 包还是不用安装。

一两个月前在另一台机器上好像是卸xorg-server 后 带X重编 fcitx5 还会报错，记不清了，也许我记错了 。
